ml-agents/mlagents/trainers/settings.py

```diff
@@ -0,0 +1,365 @@
+        )
+    # Apply cattr structure to the values
+    return cattr.structure(value, attr_fields_dict[key].type)
+
+
+def strict_to_cls(d: Mapping, t: type) -> Any:
+    if d is None:
+        return None
+    d_copy: Dict[str, Any] = {}
+    d_copy.update(d)
+    for key, val in d_copy.items():
+        d_copy[key] = check_and_structure(key, val, t)
+    return t(**d_copy)
+
+
+def trainer_settings_to_cls(d: Mapping, t: type) -> Any:
+    if d is None:
+        return None
+    d_copy: Dict[str, Any] = {}
+    d_copy.update(d)
+
+    for key, val in d_copy.items():
+        if attr.has(type(val)):
+            # Don't convert already-converted attrs classes.
+            continue
+        if key == "hyperparameters":
+            if "trainer_type" not in d_copy:
+                raise TrainerConfigError(
+                    "Hyperparameters were specified but no trainer_type was given."
+                )
+
+            else:
+                d_copy[key] = strict_to_cls(
+                    d_copy[key], TrainerSettings.to_settings(d_copy["trainer_type"])
+                )
+        elif key == "reward_signals":
+            d_copy[key] = rewardsignal_settings_to_cls(val)
+        elif key == "max_steps":
+            d_copy[key] = int(
+                float(val)
+            )  # In some configs, max steps was specified as a float
+        else:
+            d_copy[key] = check_and_structure(key, val, t)
+    return t(**d_copy)
+
+
+def rewardsignal_settings_to_cls(d: Mapping) -> Any:
+    if d is None:
+        return None
+    d_final: Dict[RewardSignalSettings.RewardSignalType, RewardSignalSettings] = {}
+
+    for key, val in d.items():
+        try:
+            enum_key = RewardSignalSettings.RewardSignalType(key)
+            t = RewardSignalSettings.to_settings(enum_key)
+            d_final[enum_key] = strict_to_cls(val, t)
+        except KeyError:
+            raise TrainerConfigError(f"Unknown reward signal type {key}")
+    return d_final
+
+
+def defaultdict_to_dict(d: DefaultDict) -> Dict:
+    return {key: cattr.unstructure(val) for key, val in d.items()}
+
+
+def dict_to_defaultdict(d: Dict, t: type) -> DefaultDict:
+    return collections.defaultdict(
+        TrainerSettings, cattr.structure(d, Dict[str, TrainerSettings])
+    )
+
+
+@attr.s(auto_attribs=True)
+class NetworkSettings:
+    @attr.s(auto_attribs=True)
+    class MemorySettings:
+        sequence_length: int = 64
+        memory_size: int = 128
+
+    normalize: bool = False
+    hidden_units: int = 128
+    num_layers: int = 2
+    vis_encode_type: EncoderType = EncoderType.SIMPLE
+    memory: Optional[MemorySettings] = None
+
+
+@attr.s(auto_attribs=True)
+class BehavioralCloningSettings:
+    demo_path: str
+    steps: int = 0
+    strength: float = 1.0
+    samples_per_update: int = 0
+    num_epoch: Optional[int] = None
+    batch_size: Optional[int] = None
```

Comment:
Could you clarify what it means for these to be `None`?  Might be worth a comment.

Reply:
For the BC module, not specifying num_epoch and batch_size means that it will use the current values for these from the trainer. This logic happens in the BC module itself - I couldn't find a good way to put the logic in `settings.py`. Will add a comment